crates/red_knot/src/types.rs

```diff
@@ -37,40 +38,137 @@ impl TypeEnvironment {
+    // name is not bound to any value
+    Unbound,
+    // TODO protocols, callable types, overloads, generics, type vars
 }
 
 impl Type {
-    pub(crate) fn name(&self) -> &str {
+    pub(crate) fn name(&self) -> Option<&str> {
         match self {
-            Type::Class(inner) => inner.name(),
-            Type::Function(inner) => inner.name(),
+            Type::Instance(inner) => Some(inner.name()),
+            Type::Function(inner) => Some(inner.name()),
+            Type::Union(_) => None,
+            Type::Intersection(_) => None,
+            Type::Any => Some("Any"),
+            Type::Never => Some("Never"),
+            Type::Unknown => Some("Unknown"),
+            Type::Unbound => Some("Unbound"),
+        }
+    }
+
+    fn display(&self, f: &mut std::fmt::Formatter, env: &TypeEnvironment) -> std::fmt::Result {
+        if let Some(name) = self.name() {
+            write!(f, "{}", name)
+        } else {
+            match self {
+                Type::Union(inner) => inner.display(f, env),
+                Type::Intersection(inner) => inner.display(f, env),
+                _ => Err(std::fmt::Error),
+            }
         }
     }
 }
 
+#[derive(Debug)]
 pub(crate) struct ClassType {
     name: Name,
 }
 
 impl ClassType {
-    pub(crate) fn name(&self) -> &str {
+    fn name(&self) -> &str {
         self.name.as_str()
     }
 }
 
+#[derive(Debug)]
 pub(crate) struct FunctionType {
     name: Name,
 }
 
 impl FunctionType {
-    pub(crate) fn name(&self) -> &str {
+    fn name(&self) -> &str {
         self.name.as_str()
     }
 }
 
+#[derive(Debug)]
+pub(crate) struct UnionType {
+    // the union type includes values in any of these types
+    elements: FxHashSet<TypeId>,
+}
+
+impl UnionType {
+    fn display(&self, f: &mut std::fmt::Formatter<'_>, env: &TypeEnvironment) -> std::fmt::Result {
+        write!(f, "(")?;
+        let mut first = true;
+        for elem_id in self.elements.iter() {
+            let ty = env.type_for_id(*elem_id);
+            if first {
+                first = false
+            } else {
+                write!(f, " | ")?;
+            };
+            ty.display(f, env)?;
+        }
+        write!(f, ")")
+    }
+}
+
+// Negation types aren't expressible in annotations, and are most likely to arise from type
+// narrowing along with intersections (e.g. `if not isinstance(...)`), so we represent them
+// directly in intersections rather than as a separate type. This sacrifices some efficiency in the
+// case where a Not appears outside an intersection (unclear when that could even happen, but we'd
+// have to represent it as a single-element intersection if it did) in exchange for better
+// efficiency in the not-within-intersection case.
+#[derive(Debug)]
+pub(crate) struct IntersectionType {
+    // the intersection type includes only values in all of these types
+    positive: FxHashSet<TypeId>,
+    // negated elements of the intersection, e.g.
+    negative: FxHashSet<TypeId>,
```

Comment:
How common are negations? We could consider boxing the `HashSet` to avoid increasing the size of all other types (this is now by far the largest type).

Reply:
I really like this idea. It doesn't really make sense to have the singleton types even belong to a module at all; they are singletons, after all. Let me explore this.

Reply:
This has a lot of interactions with my other WIP on per-module arena instead of global arena, so I'm going to explore it in the next PR instead of in this one.

Reply:
https://github.com/astral-sh/ruff/pull/11152